shell/browser/api/electron_api_utility_process.cc

```diff
@@ -0,0 +1,611 @@
+  // Currently utility process is not sandboxed which
+  // means Zygote is not used on linux, refs
+  // content::UtilitySandboxedProcessLauncherDelegate::GetZygote.
+  // If sandbox feature is enabled for the utility process, then the
+  // process reap should be signaled through the zygote via
+  // content::ZygoteCommunication::EnsureProcessTerminated.
+  base::EnsureProcessTerminated(process.Duplicate());
+  return result;
+}
+
+v8::Local<v8::Value> UtilityProcessWrapper::GetOSProcessId(
+    v8::Isolate* isolate) const {
+  if (pid_ == base::kNullProcessId)
+    return v8::Undefined(isolate);
+  return gin::ConvertToV8(isolate, pid_);
+}
+
+void UtilityProcessWrapper::ReceivePostMessage(
+    blink::TransferableMessage message) {
+  v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Value> message_value =
+      electron::DeserializeV8Value(isolate, message);
+  Emit("message", message_value);
+}
+
+void UtilityProcessWrapper::HandleMessage(ReaderType type,
+                                          std::vector<uint8_t> message) {
+  v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  auto array_buffer = v8::ArrayBuffer::New(isolate, message.size());
+  auto backing_store = array_buffer->GetBackingStore();
+  memcpy(backing_store->Data(), message.data(), message.size());
+  if (type == ReaderType::STDOUT) {
+    Emit("stdout", array_buffer,
+         base::BindRepeating(&UtilityProcessWrapper::ResumeReading,
+                             weak_factory_.GetWeakPtr(), stdout_reader_.get()));
+  } else {
+    Emit("stderr", array_buffer,
+         base::BindRepeating(&UtilityProcessWrapper::ResumeReading,
+                             weak_factory_.GetWeakPtr(), stderr_reader_.get()));
+  }
+}
+
+void UtilityProcessWrapper::ResumeReading(PipeReaderBase* pipe_io) {
+  if (!pipe_io->is_shutting_down()) {
+    pipe_io->StartReadLoop();
+  }
+}
+
+void UtilityProcessWrapper::ShutdownReader(ReaderType type) {
+  if (type == ReaderType::STDOUT) {
+    PipeIOBase::Shutdown(std::move(stdout_reader_));
+  } else {
+    PipeIOBase::Shutdown(std::move(stderr_reader_));
+  }
+}
+
+// static
+gin::Handle<UtilityProcessWrapper> UtilityProcessWrapper::Create(
+    gin::Arguments* args) {
+  gin_helper::Dictionary opts;
+  if (!args->GetNext(&opts)) {
+    args->ThrowTypeError("Expected a dictionary");
+    return gin::Handle<UtilityProcessWrapper>();
+  }
+  node::mojom::NodeServiceParamsPtr params =
+      node::mojom::NodeServiceParams::New();
+  opts.Get("modulePath", &params->script);
+  opts.Get("args", &params->args);
+  std::map<std::string, std::string> env_map;
+  if (opts.Get("env", &env_map)) {
+    for (auto const& env : env_map) {
+      params->environment.push_back(
+          node::mojom::EnvironmentVariable::New(env.first, env.second));
+    }
+  }
+  opts.Get("execArgv", &params->exec_args);
```

Comment:
we should probably throw for some (all?) of these options if the option exists but can't be converted, e.g.
```suggestion
  if (opts.Has("execArgv") && !opts.Get("execArgv", &params->exec_args)) {
    args->ThrowTypeError("Invalid value for execArgv");
    return;
  }
```